src/fswatch_win/fswatch_win_stubs.c

```diff
@@ -0,0 +1,305 @@
+/* File-watching support under Windows
+
+   We implement file-watching support under Windows using ReadDirectoryChangesW
+   and I/O completion ports. The notification of events is pull-based (one needs
+   to retrive them explicitly, there is no callback that is invoked on each
+   notification).
+
+   The main type is [fsenv] which keeps track of the state of a file watcher
+   (including all changes that have been notified by the operator system, but
+   that have not yet been retrieved by the user).
+
+   There is one value of type [wenv] for each directory being watched. It keeps
+   track of the various buffers needed to use the ReadDirectoryChangesW function
+   call.
+
+   When the first watch is created a native (not OCaml) thread is triggered
+   which runs the main event loop, waiting for notifications from the I/O
+   completion port. When a notification arrives, it is recorded in the [fsenv]
+   and it will stay there until the user retrives it by calling [get_events].
+   We notify the user that new notifications have arrived by signalling an event
+   object which causes the [wait] function to return.
+
+   Available notifications can be retrieved by calling [get_events]. The
+   [.events] field of the [fsenv] type which is mutated from multiple threads.
+   Standard CAS techniques are used to avoid data races.
+ */
+
+#include <caml/mlvalues.h>
+#include <caml/alloc.h>
+#include <caml/fail.h>
+#include <caml/osdeps.h>
+#include <caml/memory.h>
+#include <caml/threads.h>
+#include <caml/unixsupport.h>
+
+#ifdef _WIN32
+
+#include <Windows.h>
+
+#define FILE_NOTIFY_BUFFER_LENGTH ((sizeof(DWORD) * 4) * 1000)
+
+struct events {
+  struct events *next;
+  char *dirpath;
+  FILE_NOTIFY_INFORMATION *info;
+};
+
+struct fsenv {
+  struct events *events;
+  HANDLE iocp;
+  HANDLE signal;
+  struct wenv *wenv;
+};
+
+struct wenv {
+  struct wenv *next;
+  char *dirpath;
+  HANDLE dir_handle;
+  OVERLAPPED *overlapped;
+  char *buffer;
+  struct fsenv *fsenv;
+};
+
+static void push_info(struct wenv *wenv) {
+  struct events *ev = malloc(sizeof(struct events));
```

Comment:
Is this freed anywhere?

Reply:
Good catch! Pushed a fix.

Reply:
Correct me if I'm wrong because I'm not 100% sure, but since you allocate with `malloc` you should free with `free` rather than `caml_stat_free`, or use `caml_stat_alloc` here.

Reply:
I agree